### ISSUE TITLE
Windows compiler invocation opens multiple console windows

### ISSUE THREAD

**Issue by Systemcluster** [mozilla/sccache#514]
This was previously reported under #16 and supposedly fixed in [this commit](https://github.com/mozilla/sccache/commit/c70dcf258fd71dd2f114930cc308d4d409314847) and again in #472 (which I presume was included in the `0.2.10` release), but the issue still persists - upon the start of a compilation process (in particular, a `cargo` invocation), multiple console windows pop up for a split second.

I'm encountering this with the following environment:
* `sccache` version `0.2.11` installed through `cargo` (it also occured with `0.2.10` and previous versions)
* Windows 10 (x64) version `10.0.18362.0`
* Running the `cargo` subcommands `run`, `build` or `install`, with the environment variable `RUSTC_WRAPPER=sccache` set and using the Rust `msvc` toolchain.

Please let me know if you need more information.

**Comment by tangmi** [comment]
@Systemcluster Are you using PowerShell? Do you mind giving your builds a shot through Command Prompt (ensuring that sccache is not already running)? I remember reading somewhere that PowerShell does not like `CreateProcessW`'s `DETACHED_PROCESS` flag...

**Comment by Systemcluster** [comment]
@tangmi I was using PowerShell, but I'm seeing the same behavior with CMD. Same with `sccache 0.2.12-alpha.0` compiled with the changes from your branch as well.

**Comment by Systemcluster** [comment]
@tangmi I tried around a bit with some more start parameters, and after some experimentation, I decided to perform some logging with some `thread::sleep` calls to make it easier to pinpoint the source. As it turns out, the console popups are happening much later than the call to `connect_or_start_server`. (The initical sccache server start seems to create a very short lasting element in the taskbar, but no focus-stealing barrage of console windows as happening later.)

As far as I can tell (the sccache source is not the easiest to permeate) the popups appear inside `run_input_output` in util.rs. Which is very unfortunate, because at that point the process creation seems to be outsourced [to the Rust libstd](https://github.com/rust-lang/rust/blob/master/src/libstd/sys/windows/process.rs#L151).

**Comment by gilescope** [comment]
Happens on a vanilla win7 cmd console also.

**Comment by Shadlock0133** [comment]
Happens also on windows 8 in msys2's mintty.

**Comment by gilescope** [comment]
And win10

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by galich** [comment]
it's a huge problem when running in Visual Studio Code + Rust Analyzer.

Rust Analyzer invokes `cargo check` frequently while editing code and it starts to show hundreds of these windows that close themselves in a moment.

i believe we can not afford DETACHED_PROCESS anymore:

>  CREATE_NO_WINDOW... is ignored if the application is... used with either CREATE_NEW_CONSOLE or DETACHED_PROCESS.

https://docs.microsoft.com/en-us/windows/win32/procthread/process-creation-flags

**Comment by captchaKing** [comment]
For those who whant to change this behavior without recompilation:
Search for C744242808060008 (on my binary it is address 0x2b32A8) and change it to C744242800060008

Don't forget to turn off sccache server beforehand (sccache.exe --stop-server). 

If interested, bytes 08060008 correspond to CREATE_NO_WINDOW(0x8000000) | CREATE_UNICODE_ENVIRONMENT(0x400) | CREATE_NEW_PROCESS_GROUP(0x200) | DETACHED_PROCESS(0x8) considering endianness

**Comment by glandium** [comment]
My understanding of the problem is that the server being detached and not having a window, when itself spawns new processes, that may create a new window for the new process. I can imagine removing DETACHED_PROCESS might fix it, but I can also see it preventing the server from spawning properly in the first place.
It seems to me a better fix would be to use CREATE_NO_WINDOW when spawning processes from the server itself. Unfortunately, libstd doesn't have flags for that, so we'd need to use CreateProcess manually.
It also seems CREATE_NO_WINDOW should be removed from src/commands.rs, because it does nothing with DETACHED_PROCESS.

**Comment by Systemcluster** [comment]
>but I can also see it preventing the server itself from spawning properly in the first place.

Can you elaborate on this? I've used my fork for a couple months now and I noticed no such issues.

**Comment by glandium** [comment]
Isn't the first sccache process blocked until the server ends if you don't use --start-server?

**Comment by glandium** [comment]
Ok, poking around, I've convinced myself that the flag that matters the most for my concerns is CREATE_NEW_PROCESS_GROUP, and that removing DETACHED_PROCESS seems fine.

**Comment by glandium** [comment]
This is fixed by the merge of #525, but I also just discovered a side effect while I was looking with process explorer, which is that there is now a conhost process attached, which there isn't with DETACHED_PROCESS. Presumably, that's what the things that were opening new windows end up attached to, which is why windows don't show up anymore.
In the end I think we'll have to go with what was hilighted in https://github.com/mozilla/sccache/issues/514#issuecomment-741557322
I'll keep this open until that's addressed, and will keep it on the radar for upcoming release.

**Comment by glandium** [comment]
> Unfortunately, libstd doesn't have flags for that, so we'd need to use CreateProcess manually.

Actually, there is https://doc.rust-lang.org/std/os/windows/process/trait.CommandExt.html#tymethod.creation_flags

**Comment by glandium** [comment]
Interestingly, add `cmd.creation_flags(CREATE_NO_WINDOW)` makes the output from rustc entirely eaten (e.g. no warnings when there are warnings, no error messages in case of failure...)

**Comment by luser** [comment]
> Actually, there is https://doc.rust-lang.org/std/os/windows/process/trait.CommandExt.html#tymethod.creation_flags

Humorously, that exists [because I implemented it](https://github.com/rust-lang/rust/issues/37827) after having to do this the hard way when originally writing the Rust port of sccache. 🙂

**Comment by glandium** [comment]
At this point, I think we can't do better than the current (new) status, but I'll keep this open to still further investigate.

**Comment by FooBarWidget** [comment]
I am on Windows Server 2022, sccache 21185d9a15cde, Rust msvc toolchain. I cannot reproduce the issue. Neither with cmd.exe, nor with PowerShell.

**Comment by FooBarWidget** [comment]
I've also tested what happens if the let sccache start the server automatically. I **still cannot reproduce the issue**.

I've also researched the history of this issue, the relevant Windows APIs, and sccache's codebase as of commit 21185d9a15c. **I believe that this issue is now solved.**

Here's how I arrived at this conclusion:

## How does the Windows console work?

Windows differentiates between console apps and GUI apps. Windows knows whether an app is a console or a GUI app based on [a marker in the executable](https://docs.microsoft.com/en-us/cpp/build/reference/subsystem-specify-subsystem?redirectedfrom=MSDN&view=msvc-170) that's set during linking.

What happens when one starts a console app:

 1. If there's no console (e.g. the app was started from Windows Explorer) then Windows automatically allocates a console.
 2. If there is already a console (e.g. the app was started from cmd.exe, PowerShell, Windows Terminal, etc.) then Windows *attaches* the app to the existing console.
 3. If the app was started with `DETACH_PROCESS` or `CREATE_NO_NEW_WINDOW`, then Windows will neither create a new console, nor attach to the existing console. The app's won't have a console.

Okay let's say a console app (A) is started with `DETACH_PROCESS`. What happens if A starts another console app B, without any special `CreateProcess()` flags? Answer: B will be started in a new console!

But let's say console app A is started with `CREATE_NO_NEW_WINDOW`. If A starts another console app B, then B *won't* have a new console. This is not well-documented on MSDN, but [I tested this today](https://stackoverflow.com/a/71364777/20816).

Note that if both `DETACH_PROCESS` and `CREATE_NO_NEW_WINDOW` are set, then [Windows ignores `CREATE_NO_NEW_WINDOW`](https://docs.microsoft.com/en-us/windows/win32/procthread/process-creation-flags).

## Why do I think the issue is now solved?

sccache appears to have two places that start console processes (and thus that can potentially make Windows create consoles):

 1. In `run_input_output()`. Since c70dcf258fd71d, this function starts processes with the `no_console()` flag, which on Windows translates to `CREATE_NO_NEW_WINDOW`.
 2. In `run_server_process()` which is for auto-starting the sccache server as a daemon. Right now, we start the server with `CREATE_NO_NEW_WINDOW`.

Since `CREATE_NO_NEW_WINDOW`'s effect applies to the child process's children too, there shouldn't be any way for consoles to pop up. I certainly haven't been able to observe this bug.

I believe this issue can now be closed.

**Comment by FooBarWidget** [comment]
I continued researching in order to further verify, or falsify, my hypothesis that the issue is now solved.

## Artificially reproducing the problem

I can't reproduce the problem. But is that because the code is now fixed, or is that because there's something in my environment that prevents the problem from occurring? In order to answer this question, I tried to *artificially* induce the problem.

`run_input_output()` did not used to set any flags, while `run_server_process()` used to set `DETACH_PROCESS` (with `CREATE_NO_WINDOW` being ignored because of that). Thus, I expect that I will see the problem if do these:

 1. Make `run_input_output()` not set any flags.
 2. Make `run_server_process()` set `DETACH_PROCESS` instead of `CREATE_NO_WINDOW`.

After having made these changes, I indeed observed the problem: during an sccache invocation (with the server being auto-started) I observed lots of little console window popups.

Thus, I conclude that the use of `CREATE_NO_WINDOW` actually solves the problem, and that it's not just my own environment that's suppressing the problem.

## Conclusion

The conclusion is still: the issue is fixed.

**Comment by mitchhentges** [comment]
Thanks for the great investigation comments, those are helpful :+1: 

I think that this ticket has shifted since its initial report:
1. Initially, multiple console windows were opened when `sccache` starts compilation processes
2. #525 was merged, which made multi-console window no longer occur
3. Glandium raised [this comment](https://github.com/mozilla/sccache/issues/514#issuecomment-743044747), where he mentioned that the issue was fixed, but now there's a bunch of extra `conhost` processes. I'm reading between the lines here and assuming that that impacts performance in a non-trivial way.
4. It sounds like the solution proposed [here](https://github.com/mozilla/sccache/issues/514#issuecomment-741557322) is ideal, but [this later comment contradicts that](https://github.com/mozilla/sccache/issues/514#issuecomment-749203025) saying that perhaps we can't do better than opening conhost processes.

So, in summary:
* An ideal solution would not open console windows, but also wouldn't start `conhost` processes in the background
* The solution for doing ^ is a bit murky, so I'll chat with glandium for clarification.

**Comment by mitchhentges** [comment]
> The solution for doing ^ is a bit murky, so I'll chat with glandium for clarification - this will probably happen on Monday, March 14th.

Glandium has confirmed that he's comfortable deferring this ticket to a future release once we narrow down if there's still any possible improvement here.

**Comment by FooBarWidget** [comment]
I've researched the conhost issue.

## What is conhost?

Conhost is a process that manages a console and its I/O. When you run cmd.exe and you see that black window — that's a console, backed by one conhost process. Each console is backed by a different conhost process.

Normally, a cmd.exe or PowerShell session is backed by one conhost process. When you run additional CLI processes from cmd.exe/PowerShell, they make use of the same console that cmd.exe/PowerShell also uses (they *inherit* the console), and so there is only one conhost process.

### Conhost and `DETACHED_PROCESS`

When you spawn a console process B with `DETACHED_PROCESS`, that process won't have a console window, and thus it won't have a conhost.

But when B spawns a console process C — without any flags — then C *will* have a console. You can see a black window popping up. It will also have a corresponding conhost. This is consistent with my explanation in https://github.com/mozilla/sccache/issues/514#issuecomment-1059816045

~~~
A
 |
 +-- B (spawned with DETACHED_PROCESS; no console; no conhost)
     |
     +-- C (no spawn flags; has a console; has a conhost)
~~~

### Conhost and `CREATE_NO_WINDOW`

When you spawn a console process B with `CREATE_NO_WINDOW`, then that process *will* have a console. It's just hidden. Thus, B *does* have a conhost.

If B spawns a console process C with `CREATE_NO_WINDOW`, then C will *also* have a hidden console, and thus a corresponding conhost. Let's call this **exhibit 1**.

~~~
A
 |
 +-- B (spawned with CREATE_NO_WINDOW; has a hidden console; has a conhost)
     |
     +-- C (spawned with CREATE_NO_WINDOW; has a hidden console; has a conhost)
~~~
_Above: exhibit 1 in action_

But if B spawns a console process C with no extra flags, then C will inherit B's hidden console. And so C won't have a conhost.

~~~
A
 |
 +-- B (spawned with CREATE_NO_WINDOW; has a hidden console; has a conhost)
     |
     +-- C (no spawn flags; reuses B's console; reuses B's conhost)
~~~

## What does this mean for sccache?

The problem is that sccache hits exhibit 1.

 1. An sccache invocation spawns its server with `CREATE_NO_WINDOW`. This results in a conhost.
 2. The server spawns processes with `CREATE_NO_WINDOW`. This also results in a conhost.

### Proposed solution

 1. **Do** spawn the sccache server with `CREATE_NO_WINDOW`.
 2. From the server, **don't** spawn processes with `CREATE_NO_WINDOW`. Just don't use any flags. Concretely, this means that in `run_input_output` we should get rid of the `no_console()` call.

There is only one potential caveat, but it's a far-fetched corner case that I don't think anybody will run into:
 * If the server was started manually, AND also specifically with the `DETACH_PROCESS` flag, then all console subprocesses that the server spawns will pop up console windows.

**Comment by FooBarWidget** [comment]
> If my understanding is correct, then with the strong proposed solution of using CREATE_NO_WINDOW, we're going to have two conhosts, right? One for the currently running sccache command, and one for the sccache daemon and its subprocesses.

Correct.

> I'm curious: is it viable to lean on DETACHED_PROCESS instead, so that we drop to one conhost? As you said, this is only viable if every descendant of the sccache daemon is created with DETACHED_PROCESS, but is that possible? I suppose that we're at risk of, say, cargo directly spawning a new subprocess (let's say process D to continue with your convention): in which case D will pop up in a new window, eh? If it were possible to use DETACHED_PROCESS for the daemon and all of its descendants, that could result in additional performance wins, perhaps.

This is not viable. `DETACHED_PROCESS` only applies to the immediate process that we spawn, not to any subprocesses that *it* may spawn. If the sccache daemon spawns rustc with `DETACH_PROCESS`, and rustc spawns the linker, then the linker will cause a console window to pop up.

The only flag that extends to subprocesses is `CREATE_NO_WINDOW`.

**Comment by mitchhentges** [comment]
Assuming no other Windows shenanigans to influence transitive descendent processes, `CREATE_NO_WINDOW` sounds like the ideal solution. Ship it 🚢 my friend 👍
